Source/Core/AudioCommon/WASAPIStream.cpp

```diff
@@ -39,177 +44,121 @@ WASAPIStream::WASAPIStream()
 
 WASAPIStream::~WASAPIStream()
 {
-  if (m_enumerator)
-    m_enumerator->Release();
-
-  if (m_need_data_event)
-    CloseHandle(m_need_data_event);
-
-  if (m_running)
-  {
-    m_running = false;
-    if (m_thread.joinable())
-      m_thread.join();
-  }
-
-  CoUninitialize();
+  m_running.store(false, std::memory_order_relaxed);
+  if (m_thread.joinable())
+    m_thread.join();
 }
 
 bool WASAPIStream::isValid()
 {
   return true;
 }
 
-static bool HandleWinAPI(std::string message, HRESULT result)
+static bool HandleWinAPI(std::string_view message, HRESULT result)
 {
-  if (result != S_OK)
+  if (FAILED(result))
   {
-    _com_error err(result);
-    std::string error = TStrToUTF8(err.ErrorMessage()).c_str();
+    std::string error;
 
     switch (result)
     {
     case AUDCLNT_E_DEVICE_IN_USE:
       error = "Audio endpoint already in use!";
       break;
+    default:
+      error = TStrToUTF8(_com_error(result).ErrorMessage()).c_str();
+      break;
     }
 
-    ERROR_LOG(AUDIO, "WASAPI: %s: %s", message.c_str(), error.c_str());
+    ERROR_LOG(AUDIO, "WASAPI: %s: %s", std::string(message).c_str(), error.c_str());
   }
 
-  return result == S_OK;
+  return SUCCEEDED(result);
 }
 
-std::vector<std::string> WASAPIStream::GetAvailableDevices()
+static void ForEachNamedDevice(std::function<bool(ComPtr<IMMDevice>, std::string)> callback)
 {
-  CoInitialize(nullptr);
+  HRESULT result = CoInitializeEx(nullptr, COINIT_MULTITHREADED);
+  // It's fine if CoInitializeEx fails with RPC_E_CHANGED_MODE
+  if (result != RPC_E_CHANGED_MODE && !HandleWinAPI("Failed to call CoInitialize", result))
+    return;
+
+  wil::unique_couninitialize_call cleanup;
+  if (FAILED(result))
+    cleanup.release();  // Only call CoUninitialize on success
```

Comment:
Kinda same reasoning as for the "It's fine" comment: _why_? Omitting that just leaves the reader with more questions than answers (at least in my case where I'm fairly rusty in COM and stuff)